### PR TITLE
Add support for the mode of fallocate

### PR DESCRIPTION
**Pull request by liqinggd**
Merge https://github.com/occlum/sefs/pull/23 first

### CODE REVIEW

File: test/file/main.c [hunk 1]
```diff
@@ -203,6 +215,236 @@ static int __test_posix_fallocate(const char *file_path) {
...
+    }
+    off_t len = 2 * BLK_SIZE;
+    if (fill_file_with_repeated_bytes(fd, len, 0xFF) < 0) {
+        THROW_ERROR("failed to fill file");
+    }
+    int mode = FALLOC_FL_KEEP_SIZE | 0xDEAD;
+    off_t offset = 0;
+    len = len / 2;
+    int ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EOPNOTSUPP) {
```
Comment (tatetian) [inline test/file/main.c:238]: I think `ret < 0 && errno == EOPNOTSUPP` is more straightforward.

File: test/file/main.c [hunk 2]
```diff
@@ -203,6 +215,236 @@ static int __test_posix_fallocate(const char *file_path) {
...
+    mode = FALLOC_FL_COLLAPSE_RANGE | FALLOC_FL_KEEP_SIZE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EINVAL ) {
+        THROW_ERROR("failed to check collapse range with other flags");
+    }
+    mode = FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE | FALLOC_FL_UNSHARE_RANGE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EINVAL ) {
+        THROW_ERROR("failed to check unshare flag without allocate mode");
+    }
```
Comment (tatetian) [inline test/file/main.c:265]: I think writing a loop that iterates an array of input arguments and expected output is more concise and readable. And also easier to add new test cases in the future.

**Review by tatetian — COMMENTED** [review]
A general advice for readability. Code would be more clear when separated in code blocks. I reviewed several PRs of you and found you are not quite used to add blank lines in functions. Most non-trivial functions you write do not have a single empty line. I think this something you can work on and improve in the future.

I also reviewed the related PR to sefs.

File: test/file/main.c (hunk 2 above)
Comment (liqinggd) [reply test/file/main.c:260]: Good suggestion

File: test/file/main.c (hunk 1 above)
Comment (liqinggd) [reply test/file/main.c]: `!(ret < 0 && errno == EOPNOTSUPP)` is equal to `ret >= 0 || errno != EOPNOTSUPP`

File: src/libos/src/fs/file_ops/fallocate.rs [hunk 3]
```diff
@@ -1,11 +1,93 @@
...
+                "Unshare range should only be used with allocate mode"
+            );
+        }
+        Ok(flags)
+    }
+}
+
+pub trait IntoFallocateMode {
+    fn into_fallocate_mode(self) -> FallocateMode;
+}
```
Comment (tatetian) [inline src/libos/src/fs/file_ops/fallocate.rs:72]: Use the standard `From` trait for this purpose by implementing `From<FallocateFlags>` for `FallocateMode`.

Comment (liqinggd) [reply src/libos/src/fs/file_ops/fallocate.rs]: Thanks for the advice, done.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**